From bf2f91f22c0146ed27560b71dd6ba25b6e8b3243 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@st.com>
Date: Fri, 18 May 2018 17:28:11 +0200
Subject: [PATCH 7/7] waylandsink: Add prototype for gst_wayand_pool_get_type()

---
 ext/wayland/gstwaylandsink.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index b70e74e..84f67e9 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -126,6 +126,9 @@ G_DEFINE_TYPE_WITH_CODE (GstWaylandSink, gst_wayland_sink, GST_TYPE_VIDEO_SINK,
  * width/height + stride and use the viewporter interface to crop, a bit like
  * we use to do with XV. It would still be quite limited. It's a bit retro,
  * hopefully there will be a better Wayland interface in the future. */
+
+GType gst_wayland_pool_get_type (void);
+
 typedef struct
 {
   GstVideoBufferPool parent;
-- 
2.7.4

