From 900646cd52939451325e11fecf6524dbde984136 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@linaro.org>
Date: Wed, 21 Jan 2015 16:48:38 +0100
Subject: [PATCH 1/2] add rule to compile the driver out of kernel tree

Signed-off-by: Benjamin Gaignard <benjamin.gaignard@linaro.org>
---
 Makefile | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/Makefile b/Makefile
index 469dae5..64393a8 100644
--- a/Makefile
+++ b/Makefile
@@ -45,3 +45,12 @@ endif
 
 optee-objs += generic/tee_debug.o
 endif # CFG_TEE_DRV_DEBUGFS
+
+ifneq ($(CFG_TEE_CORE_ADDR),)
+ccflags-y+=-DCFG_TEE_CORE_ADDR=${CFG_TEE_CORE_ADDR}
+else
+$(error *** you must specify the address CFG_TEE_CORE_ADDR *** )
+endif
+
+all:
+	$(MAKE) ARCH=$(ARCH) -C $(KDIR) M=$(CURDIR) modules CFG_TEE_CORE_ADDR=$(CFG_TEE_CORE_ADDR)
-- 
1.9.1

