From f314a0e2c6c753edea82e65bde5f7fbba0f092c1 Mon Sep 17 00:00:00 2001
From: Bryce Harrington <bryce@osg.samsung.com>
Date: Thu, 22 Sep 2016 10:33:46 -0700
Subject: [PATCH 01/61] configure.ac: bump version to 1.12.90 for open
 development

Signed-off-by: Bryce Harrington <bryce@osg.samsung.com>
---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 619d3f6..6398b2a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,6 +1,6 @@
 m4_define([weston_major_version],  [1])
 m4_define([weston_minor_version],  [12])
-m4_define([weston_micro_version],  [0])
+m4_define([weston_micro_version],  [90])
 m4_define([weston_version],
           [weston_major_version.weston_minor_version.weston_micro_version])
 m4_define([libweston_major_version], [1])
-- 
2.7.4

