From 66a161460662d083cfe258875d49318f43183a17 Mon Sep 17 00:00:00 2001
From: Khem Raj <raj.khem@gmail.com>
Date: Thu, 7 Jan 2016 19:33:53 +0000
Subject: [PATCH 05/19] Add gnu source

---
 testcases/kernel/connectors/pec/pec_listener.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/testcases/kernel/connectors/pec/pec_listener.c b/testcases/kernel/connectors/pec/pec_listener.c
index d11e4dc..b8b5e0d 100644
--- a/testcases/kernel/connectors/pec/pec_listener.c
+++ b/testcases/kernel/connectors/pec/pec_listener.c
@@ -20,6 +20,7 @@
 /*                                                                            */
 /******************************************************************************/
 
+#define _GNU_SOURCE
 #include <sys/socket.h>
 #include <sys/poll.h>
 #include <sys/types.h>
-- 
2.7.4

